csrc/ir/nodes.cpp

```diff
@@ -4371,7 +4372,17 @@ std::vector<PolymorphicValue> MatmulOp::evaluate(
     const std::vector<PolymorphicValue>& inputs) const {
   const auto a = inputs.at(0).as<at::Tensor>();
   const auto b = inputs.at(1).as<at::Tensor>();
-  return {at::matmul(a, b)};
+
+  auto matmul_out = at::matmul(a, b);
```

Comment:
at::matmul_out is not used since it does not allow inputs/outputs which require gradients.
https://github.com/pytorch/pytorch/blob/1f3d8896bc9cea7f46c50ff92b69c6aa139defcb/aten/src/ATen/native/LinearAlgebra.cpp#L2018-L2025

Reply:
This is suspicious -- We are not using ExpressionEvaluator to build a DAG for autograd, so inputs/outputs here shouldn't require grads. Where did inputs/outputs get requires_grads? Your test case didn't `torch.randn(..., requires_grad=True)` obviously to start with. 

cc @jjsjann123

Reply:
You might be right, the tensor evaluations themselves will likely not have the requires_grad field.
I had ruled out `at::matmul_out` after going through its code, and trying it independently that raised this condition, which may have been premature.

Reply:
I attempted a complete example in nvfuser and I do get an error.
I need to dig into the expression evaluator on how this flag is propagated/inferred. Expression evaluator will have this information for the fusion inputs.

Reply:
I'm fine with addressing this in a separate PR. But I'd still try to understand the requires_grad bit sooner than later -- it shouldn't have been there and it blocks optimization to matmul into a pre-allocated output.

Reply:
Maybe we can try this https://github.com/pytorch/pytorch/blob/a3688ead4bcd13de7d81a9e114b1c8cfc332970a/c10/core/GradMode.h#L31-L33

Reply:
but yeah, we can do it in a separate PR. I'm more than happy to review that.